### PR TITLE
[MRG] add tangent space with euclidean mean.

### PR DESCRIPTION
**Pull request by GaelVaroquaux**
Geometric mean can be unstable, and not converge. It makes it hazardous
to use tangent-space parametrization (feedback at OHBM). A Euclidean
mean works well, and this PR adds an option to have tangent-space
parametrization with Euclidean mean.

### CODE REVIEW

File: nilearn/connectome/connectivity_matrices.py [hunk 1]
```diff
@@ -396,8 +396,11 @@ class ConnectivityMeasure(BaseEstimator, TransformerMixin):
         towards zero compared to a maximum-likelihood estimate
 
     kind : {"correlation", "partial correlation", "tangent",\
-            "covariance", "precision"}, optional
-        The matrix kind.
+            "tangent_mean", "covariance", "precision"}, optional
+        The matrix kind. "tangent" refers to transforming the data to
+        tangent reprensentation of the differences to the mean of
+        matrices. "tangent_mean" uses in addition a geometric mean to
```
Comment (bthirion) [inline nilearn/connectome/connectivity_matrices.py]: 'tangent_geometric' ?

**Comment by bthirion** [comment]
LGTM : just a naming comment.

**Comment by bthirion** [comment]
+1 for merge.

File: nilearn/connectome/connectivity_matrices.py [hunk 2]
```diff
@@ -396,8 +396,11 @@ class ConnectivityMeasure(BaseEstimator, TransformerMixin):
         towards zero compared to a maximum-likelihood estimate
 
     kind : {"correlation", "partial correlation", "tangent",\
-            "covariance", "precision"}, optional
-        The matrix kind.
+            "tangent_geometric", "covariance", "precision"}, optional
+        The matrix kind. "tangent" refers to transforming the data to
+        tangent reprensentation of the differences to the mean of
```
Comment (KamalakerDadi) [inline nilearn/connectome/connectivity_matrices.py]: typo here

File: nilearn/connectome/connectivity_matrices.py [hunk 3]
```diff
@@ -396,8 +396,11 @@ class ConnectivityMeasure(BaseEstimator, TransformerMixin):
         towards zero compared to a maximum-likelihood estimate
 
     kind : {"correlation", "partial correlation", "tangent",\
-            "covariance", "precision"}, optional
-        The matrix kind.
+            "tangent_geometric", "covariance", "precision"}, optional
```
Comment (KamalakerDadi) [inline nilearn/connectome/connectivity_matrices.py]: While we are at it, can we write explicit naming to tangent too ?

**Comment by GaelVaroquaux** [comment]
> While we are at it, can we write explicit naming to tangent too ?

What do you mean?

**Comment by KamalakerDadi** [comment]
- adding in whats_new.rst is missing. I think it is useful.
- could you please add tests ?

**Comment by GaelVaroquaux** [comment]
>   • adding in whats_new.rst is missing. I think it is useful.

Yes.

>   • could you please add tests ?

OK, but I fear that they will be just smoke test

**Comment by KamalakerDadi** [comment]
>While we are at it, can we write explicit naming to tangent too ?
>What do you mean?

tangent_euclidean ?

**Comment by GaelVaroquaux** [comment]
> tangent_euclidean ?

I'd rather not: let's keep it simple. This terminology already frightens
our users too much.

File: nilearn/connectome/connectivity_matrices.py [hunk 4]
```diff
@@ -466,20 +455,55 @@ def fit(self, X, y=None):
...
         return self
 
+    def fit_transform(self, X, y=None):
+        if self.kind in ('tangent', 'tangent_geometric'):
+            # Check that people are applying fit_transform to a group of
+            # subject
+            # We can only impose this in fit_transform, as it is legit to
+            # fit only on a single given reference point
+            if not len(X) > 1:
+                raise ValueError("Tangent space parametrization can only "
```
Comment (KamalakerDadi) [inline nilearn/connectome/connectivity_matrices.py]: Can you add a test for this too. To test if error raises ?

**Comment by KamalakerDadi** [comment]
May be you have missed to add 'tangent_geometric' at Line552 ?
Currently, there is a chance to override mean_ ? Correct me if I am wrong

**Comment by GaelVaroquaux** [comment]
Thanks a lot for your comments @KamalakerDadi, in particular for asking for tests. Indeed, writing tests, I uncovered a set of bugs, which I just fixed.

File: nilearn/connectome/connectivity_matrices.py [hunk 5]
```diff
@@ -584,6 +616,13 @@ def inverse_transform(self, connectivities, diagonal=None):
 
         if self.kind == 'tangent':
             mean_sqrt = _map_eigenvalues(lambda x: np.sqrt(x), self.mean_)
+            idty = np.eye(self.mean_.shape[0])
+            connectivities = [mean_sqrt.dot(
+                displacement + idty).dot(mean_sqrt)
```
Comment (bthirion) [inline nilearn/connectome/connectivity_matrices.py]: Shouldn't one of the two  mean_sqrt be transposed ? (not sure that they are symmetric)

**Comment by bthirion** [comment]
LGTM besides the small detail.

**Comment by KamalakerDadi** [comment]
>LGTM besides the small detail.

Currently, I am using this PR in my experimental paper. To see the results and also one way reviewing this PR.

File: nilearn/connectome/connectivity_matrices.py [hunk 6]
```diff
@@ -466,20 +455,59 @@ def fit(self, X, y=None):
...
+        self._check_input(X)
+        self.cov_estimator_ = clone(self.cov_estimator)
+
+        covariances = [self.cov_estimator_.fit(x).covariance_ for x in X]
+        if self.kind == 'tangent_geometric':
             self.mean_ = _geometric_mean(covariances, max_iter=30, tol=1e-7)
+        else:
+            self.mean_ = np.mean(covariances, axis=0)
+            # Fight numerical instabilities: make symmetric
+            self.mean_ = self.mean_ + self.mean_.T
```
Comment (KamalakerDadi) [inline nilearn/connectome/connectivity_matrices.py]: If I understood correctly, this also applies for all kinds except tangent_geometric ?

**Comment by salma1601** [comment]
Why not just switch back to euclidean mean in case the algorithm does not converge? We can tell this in the actual non convergence warning. I find that this multiplication of kinds may complicate users life.

**Comment by bthirion** [comment]
You probably don't want to switch model while dealing with convergence issues. This makes the behavior of estimators even more complex to figure out.
My 2c,
B

**Comment by GaelVaroquaux** [comment]
> Why not just switch back to euclidean mean in case the algorithm does
> not converge? We can tell this in the actual non convergence warning. I
> find that this multiplication of kinds may complicate users life.

Kamalakar is running benchmarks right now to figure out those tradeoffs
based on empirical results.

G

**Comment by salma1601** [comment]
>Kamalakar is running benchmarks right now to figure out those tradeoffs based on empirical results.

Great ! The gradient descent is initiated at the euclidean mean, so we'll see if it can be trapped in an aberrant region.

**Comment by kchawla-pi** [comment]
Is this alive? If not, please close it by this weekend.

**Comment by emdupre** [comment]
@GaelVaroquaux confirmed this has been superseded. Closing :door:

**Comment by kchawla-pi** [comment]
Thanks! @emdupre
